### PR TITLE
replica_aggregator stores replicas

### PR DESCRIPTION
**Pull request by kenoir**
Allow the replica_aggregator to store replicas, always succeeds.

For https://github.com/wellcometrust/platform/issues/3501

### CODE REVIEW

File: replica_aggregator/src/main/resources/application.conf [hunk 1]
```diff
@@ -3,3 +3,4 @@ aws.metrics.namespace=${?metrics_namespace}
 aws.outgoing.sns.topic.arn=${?outgoing_topic_arn}
 aws.sqs.queue.url=${?queue_url}
 operation.name=${?operation_name}
+aws.replicas.dynamo.tableName=${?replicas_table_name}
```
Comment (alexwlchan) [inline replica_aggregator/src/main/resources/application.conf]: nit: I'm trying to keep these alphabetised, to make it easier to read

File: replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/DynamoVersionedStore.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package uk.ac.wellcome.platform.storage.replica_aggregator
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
+import org.scanamo.DynamoFormat
+import uk.ac.wellcome.storage.dynamo.{DynamoConfig, DynamoHashRangeEntry}
+import uk.ac.wellcome.storage.store.VersionedStore
+import uk.ac.wellcome.storage.store.dynamo.DynamoHashRangeStore
+
+// TODO: Should be in storage library
```
Comment (alexwlchan) [inline replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/DynamoVersionedStore.scala]: Create a GitHub ticket for it?

File: replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/models/ReplicaPath.scala [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package uk.ac.wellcome.platform.storage.replica_aggregator.models
 
-class ReplicaPath(val underlying: String) extends AnyVal
+case class ReplicaPath(val underlying: String) extends AnyVal {
+  override def toString: String = underlying
+}
```
Comment (alexwlchan) [inline replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/models/ReplicaPath.scala]: If you've done this I think you can ditch the `object` below?

File: replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregator.scala [hunk 4]
```diff
@@ -2,15 +2,16 @@ package uk.ac.wellcome.platform.storage.replica_aggregator.services
...
 import uk.ac.wellcome.storage.store.VersionedStore
 import uk.ac.wellcome.storage.{Identified, Version}
 
 import scala.util.Try
 
 class ReplicaAggregator(
-  versionedStore: VersionedStore[ReplicaPath, Int, Set[ReplicaResult]]
-) {
+  versionedStore: VersionedStore[ReplicaPath, Int, List[ReplicaResult]]
+) extends Logging {
```
Comment (alexwlchan) [inline replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregator.scala]: We don't seem to be logging anything.

File: replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregatorWorker.scala [hunk 5]
```diff
@@ -35,26 +35,24 @@ class ReplicaAggregatorWorker[IngestDestination, OutgoingDestination](
...
-      replicationSet = replicationSet,
-      startTime = Instant.now,
-      endTime = Instant.now
-    )
+      // TODO: Need to distinguish result to determine outgoing message:
+      // ReplicationAggregationIncomplete/ReplicationAggregationComplete
+      // How does that map to IngestStepSucceeded & then sendIfSuccessful?
 
-    val ingestStep = IngestStepSucceeded(summary)
+      ingestStep <- Success(aggregation match {
```
Comment (alexwlchan) [inline replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregatorWorker.scala]: Do you need to wrap this in a `Success` and a map? Could you do:

```scala
ingestStep = aggregation match {
   ...
}
```

File: replica_aggregator/src/test/scala/uk/ac/wellcome/platform/storage/replica_aggregator/ReplicaAggregatorFeatureTest.scala [hunk 6]
```diff
@@ -42,6 +55,30 @@ class ReplicaAggregatorFeatureTest
...
+          val replicaResult = stored.identifiedT.head
+
+          inside(replicaResult) {
+            case ReplicaResult(
+                ingestId,
+                PrimaryStorageLocation(
+                  storageProvider,
+                  location
+                ),
+                _
```
Comment (alexwlchan) [inline replica_aggregator/src/test/scala/uk/ac/wellcome/platform/storage/replica_aggregator/ReplicaAggregatorFeatureTest.scala]: Maybe assert that the replica result time is recent?

File: terraform/stack/dynamo.tf [hunk 7]
```diff
@@ -15,9 +15,50 @@ module "replicator_lock_table" {
 locals {
   versioner_versions_table_name  = "${aws_dynamodb_table.versioner_versions_table.name}"
   versioner_versions_table_index = "ingestId_index"
+
+  replicas_table_name = "${aws_dynamodb_table.replicas_table.name}"
+}
+
+# TODO: Move this into the 'critical' part
```
Comment (alexwlchan) [inline terraform/stack/dynamo.tf]: Related: https://github.com/wellcometrust/platform/issues/3722

File: terraform/stack/dynamo.tf [hunk 8]
```diff
@@ -15,9 +15,50 @@ module "replicator_lock_table" {
...
+    name = "version"
+    type = "N"
+  }
 }
 
-# TODO: MOve this into the 'critical' part
+data "aws_iam_policy_document" "replicas_table_readwrite" {
+  statement {
+    actions = [
+      "dynamodb:DeleteItem",
```
Comment (alexwlchan) [inline terraform/stack/dynamo.tf]: Why do we need to grant DeleteItem privileges? Or UpdateItem? I would think the VersionedStore just needs Get, Query and Put.

File: replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregatorWorker.scala (hunk 5 above)
Comment (kenoir) [reply replica_aggregator/src/main/scala/uk/ac/wellcome/platform/storage/replica_aggregator/services/ReplicaAggregatorWorker.scala]: I think so. `IngestFailed` & `IngestStepSucceeded` are both rightly `Success` for the `processMessage` method.

File: replica_aggregator/src/test/scala/uk/ac/wellcome/platform/storage/replica_aggregator/ReplicaAggregatorFeatureTest.scala (hunk 6 above)
Comment (kenoir) [reply replica_aggregator/src/test/scala/uk/ac/wellcome/platform/storage/replica_aggregator/ReplicaAggregatorFeatureTest.scala]: I feel that's a bit too fuzzy and ambiguity here might lead to confusion later. 

In that vein I do think it is worth asserting something about that value and not simply `_` it out (i'm going to assert it is an Instant for now).

**Pull request merged**